rest/src/main/java/com/redhat/ipaas/rest/v1/dao/jsondb/JsonDBDAO.java

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.redhat.ipaas.rest.v1.dao.jsondb;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.MapType;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import com.redhat.ipaas.rest.v1.controller.handler.exception.IPaasServerException;
+import com.redhat.ipaas.jsondb.JsonDB;
+import com.redhat.ipaas.rest.v1.dao.DataAccessObject;
+import com.redhat.ipaas.rest.v1.model.ListResult;
+import com.redhat.ipaas.rest.v1.model.WithId;
+import com.redhat.ipaas.rest.v1.util.Json;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+
+/**
+ * Implements a DataAccessObject using the {@see: JsonDB}.
+ */
+abstract public class JsonDBDAO<T extends WithId> implements DataAccessObject<T> {
+
+    private final JsonDB rtdb;
+
+    public JsonDBDAO(JsonDB rtdb) {
+        this.rtdb = rtdb;
+    }
+
+    abstract public String getCollectionPath();
+
+    @Override
+    public T fetch(String id) {
+        try {
+            String dbPath = getCollectionPath()+"/:"+id;
+            byte[] json = rtdb.getAsByteArray(dbPath);
+            if( json==null || json.length == 0 ) {
+                return null;
+            }
+            return Json.mapper().readValue(json, getType());
+        } catch (RuntimeException|IOException e) {
+            throw IPaasServerException.launderThrowable(e);
+        }
+    }
+
+    @Override
+    public ListResult<T> fetchAll() {
+        try {
+            // get the data out..
+            byte[] json = rtdb.getAsByteArray(getCollectionPath());
+            if( json!=null && json.length > 0 ) {
+
+                // Lets use jackson to parse the map of keys to our model instances
+                ObjectMapper mapper = Json.mapper();
+                TypeFactory typeFactory = mapper.getTypeFactory();
+                MapType mapType = typeFactory.constructMapType(LinkedHashMap.class, String.class, getType());
+                LinkedHashMap<String, T> map = mapper.readValue(json, mapType);
+
+                return ListResult.of(map.values());
+            } else {
+                return ListResult.of(Collections.EMPTY_LIST);
+            }
+        } catch (RuntimeException|IOException e) {
+            throw IPaasServerException.launderThrowable(e);
+        }
+    }
+
+    @Override
+    public T create(T entity) {
+        try {
+
+            String dbPath = getCollectionPath()+"/:"+entity.getId().get();
+
+            // Only create if it did not exist.
+            if( rtdb.exists(dbPath) ) {
+                return null;
+            }
+
+            byte[] json = Json.mapper().writeValueAsBytes(entity);
```

Comment:
Please correct me, but if I understand the approach of JsonDB right then the client must include all related objects in the request, too. For example in the current model, there's a `Connector.getConnectorGroup()` and this group has to be fully provided, so when a ConnectorGroup's field changes all Connector documents must be touched to update this in the `Connector` documents, right ?

Also, what happens with circular dependencies (e.g. when there would be a `ConnectorGroup.getConnectors()` for bidirectional navigation) ? A real existing example is at the moment `Connection.getTags()` and `Tag.getConnections()`.

I think mapping relationships that way and especially updates of fields (like global configuration values) require a complete scan through all documents to update the included values. And there's of course the possibility of inconsistencies here, but that's not so uncommon in the NoSQL world ;)

It also puts quite a lot of burden to the client to provide always all dependents objects in their full beauty when creating objects.